// electron-builder/out/targets/appImage.js
"use strict";

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

const platformPackager_1 = require("../platformPackager");
const metadata_1 = require("../metadata");
const path = require("path");
const util_1 = require("../util/util");
const fs_extra_p_1 = require("fs-extra-p");
const binDownload_1 = require("../util/binDownload");
const bluebird_1 = require("bluebird");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("../util/awaiter");
const appImageVersion = "AppImage-5";
//noinspection SpellCheckingInspection
const appImagePathPromise = binDownload_1.getBin("AppImage", appImageVersion, `https://dl.bintray.com/electron-userland/bin/${ appImageVersion }.7z`, "19833e5db3cbc546432de8ddc8a54181489e6faad4944bd1f3138adf4b771259");
class AppImageTarget extends platformPackager_1.TargetEx {
    constructor(packager, helper, outDir) {
        super("appImage");
        this.packager = packager;
        this.helper = helper;
        this.outDir = outDir;
        this.desktopEntry = helper.computeDesktopEntry("AppRun", `X-AppImage-Version=${ packager.appInfo.buildVersion }`);
    }
    build(appOutDir, arch) {
        return __awaiter(this, void 0, void 0, function* () {
            const packager = this.packager;
            const image = path.join(this.outDir, packager.generateName("AppImage", arch, false));
            const appInfo = packager.appInfo;
            yield util_1.unlinkIfExists(image);
            const appImagePath = yield appImagePathPromise;
            const args = ["-joliet", "on", "-volid", "AppImage", "-dev", image, "-padding", "0", "-map", appOutDir, "/usr/bin", "-map", path.join(__dirname, "..", "..", "templates", "linux", "AppRun.sh"), `/AppRun`, "-map", yield this.desktopEntry, `/${ appInfo.name }.desktop`, "-move", `/usr/bin/${ appInfo.productFilename }`, `/usr/bin/${ appInfo.name }`];
            for (let _ref of yield this.helper.icons) {
                var _ref2 = _slicedToArray(_ref, 2);

                let from = _ref2[0];
                let to = _ref2[1];

                args.push("-map", from, `/usr/share/icons/default/${ to }`);
            }
            // must be after this.helper.icons call
            if (this.helper.maxIconPath == null) {
                throw new Error("Icon is not provided");
            }
            args.push("-map", this.helper.maxIconPath, "/.DirIcon");
            args.push("-chown_r", "0", "/", "--");
            args.push("-zisofs", `level=${ packager.devMetadata.build.compression === "store" ? "0" : "9" }:block_size=128k:by_magic=off`);
            args.push("set_filter_r", "--zisofs", "/");
            yield util_1.exec(process.platform === "darwin" ? path.join(appImagePath, "xorriso") : "xorriso", args);
            yield new bluebird_1.Promise((resolve, reject) => {
                const rd = fs_extra_p_1.createReadStream(path.join(appImagePath, arch === metadata_1.Arch.ia32 ? "32" : "64", "runtime"));
                rd.on("error", reject);
                const wr = fs_extra_p_1.createWriteStream(image, { flags: "r+" });
                wr.on("error", reject);
                wr.on("finish", resolve);
                rd.pipe(wr);
            });
            const fd = yield fs_extra_p_1.open(image, "r+");
            try {
                const magicData = new Buffer([0x41, 0x49, 0x01]);
                yield fs_extra_p_1.write(fd, magicData, 0, magicData.length, 8);
            } finally {
                yield fs_extra_p_1.close(fd);
            }
            yield fs_extra_p_1.chmod(image, "0755");
            packager.dispatchArtifactCreated(image, packager.generateName("AppImage", arch, true));
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AppImageTarget;
